### PR TITLE
Fixed the zone-based loss function.

### PR DESCRIPTION
**Pull request by mhrmm**
@zhengyaogu Some small bugfixes here and there.

### CODE REVIEW

**Comment by zhengyaogu** [comment]
I have some concerns about the change to loss.py. Yesterday, when I was training the model, I found that clamping actually yield worse accuracies. A more direct way to prevent exploding gradient is to simply clamp the gradient. But this is difficult to implement. Based on yesterday's experiment, the training accuracy progressed steadily, showing no signs of exploding gradients.

I will keep this pr open and train the network with the new loss function, experimenting with different clamping thresholds, to find the optimal configuration. After that, I will close this pr.

**Comment by zhengyaogu** [comment]
I update this branch in another pull request. I am closing this issue
